File: Backtracking/47_PermutationsII.cpp
/*
 * @Author: xuezaigds@gmail.com
 * @Last Modified time: 2016-04-20 10:38:44
 */

class Solution {
public:
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        vector<vector<int>> ans;
        sort(nums.begin(), nums.end());
        helper(nums, 0, ans);
        return ans;
    }

    void helper(vector<int> num, int begin, vector<vector<int>> &ans){
        // num is passed by value
        if(begin==num.size()-1){
            ans.push_back(num);
            return;
        }
        for(int i=begin;i<num.size();i++){
            if(i!=begin && num[i]==num[begin])  continue;
            swap(num[i], num[begin]);
            helper(num, begin+1, ans);
            // No Backtracking here!!!
        }
    }
};


class Solution_2{
public:
    /*
     * 1. sort nums in ascending order, add it to res;
     * 2. generate the next permutation of nums, and add it to res;
     * 3. repeat 2 until the next permutation of nums returns to the sorted condition in 1.
     */
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        vector<vector<int>> ans;
        sort(nums.begin(), nums.end());
        ans.push_back(nums);
        while(nextPermutation(nums)){
            ans.push_back(nums);
        }
        return ans;
    }

    bool nextPermutation(vector<int>& nums) {
        if(nums.empty())    return false;
        int i;
        for(i=nums.size()-1; i>=1; i--){
            if(nums[i] > nums[i-1]) break;
        }
        reverse(nums.begin()+i, nums.end());
        if(i==0)    return false;
        auto iter = upper_bound(nums.begin()+i, nums.end(), nums[i-1]);
        swap(nums[i-1], *iter);
        return true;
    }
};

/*
[]
[1]
[1,2,3]
[2,2,3,3]
*/

File: Backtracking/47_PermutationsII.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-


class Solution(object):
    # Easy to understand: recursively.
    # Just like get permute for distinct numbers.
    def permuteUnique(self, nums):
        ans = []
        nums.sort()
        self.dfs(nums, 0, ans)
        return ans

    def dfs(self, num, begin, ans):
        if begin == len(num) - 1:
            ans.append(num)
            return

        for i in range(begin, len(num)):
            if i != begin and num[i] == num[begin]:
                continue
            num[i], num[begin] = num[begin], num[i]
            # num[:], get a new copy.  Just like pass by value
            self.dfs(num[:], begin + 1, ans)


class Solution_2(object):
    '''
    1. sort nums in ascending order, add it to res;
    2. generate the next permutation of nums, and add it to res;
    3. repeat 2 until the next permutation of nums.
    '''
    def permuteUnique(self, nums):
        nums.sort()
        ans = []
        ans.append(nums[:])
        while self.nextPermutation(nums):
            ans.append(nums[:])

        return ans

    def nextPermutation(self, nums):
        length = len(nums)
        index = length - 1

        while index >= 1:
            if nums[index] > nums[index - 1]:
                for i in range(length - 1, index - 1, -1):
                    if nums[i] > nums[index - 1]:
                        nums[i], nums[index - 1] = nums[index - 1], nums[i]
                        nums[index:] = sorted(nums[index:])
                        return True
            else:
                index -= 1

        # Nums is in descending order, just reverse it.
        return False


"""
[]
[1]
[1,2,3]
[2,2,3,3]
"""
